arangod/RestServer/FlushFeature.cpp

```diff
@@ -20,20 +20,421 @@
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("failed to find handler while applying 'Flush' recovery marker of type '") + type + "'"
+    );
+  }
+
+  TRI_ASSERT(itr->second); // non-nullptr ensured by registerFlushRecoveryCallback(...)
+
+  auto* dbFeature = arangodb::application_features::ApplicationServer::lookupFeature<
+    arangodb::DatabaseFeature
+  >("Database");
+
+  if (!dbFeature) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("failure to find feature 'Database' while applying 'Flush' recovery marker of type '") + type + "'"
+    );
+  }
+
+  auto* vocbase = dbFeature->useDatabase(dbId);
+
+  if (!vocbase) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("failed to find database '") + std::to_string(dbId) + "' while applying 'Flush' recovery marker of type '" + type + "'"
+    );
+  }
+
+  TRI_DEFER(vocbase->release());
+
+  if (!slice.hasKey(TYPE_ATTRIBUTE)) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("failed to find a 'data' attribute inside the supplied marker body while applying 'Flush' recovery marker of type '") + type + "'"
+    );
+  }
+
+  auto data = slice.get(DATA_ATTRIBUTE);
+
+  try {
+    return itr->second(*vocbase, data);
+  } catch (arangodb::basics::Exception const& e) {
+    return arangodb::Result(
+      e.code(),
+      std::string("caught exception while applying 'Flush' recovery marker of type '") + type + "': " + e.what()
+    );
+  } catch (std::exception const& e) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("caught exception while applying 'Flush' recovery marker of type '") + type + "': " + e.what()
+    );
+  } catch (...) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      std::string("caught exception while applying 'Flush' recovery marker of type '") + type + "'"
+    );
+  }
+}
+
+class MMFilesFlushMarker final: public arangodb::MMFilesWalMarker {
+ public:
+  /// @brief read constructor
+  MMFilesFlushMarker(MMFilesMarker const& marker) {
+    TRI_ASSERT(type() == marker.getType());
+    auto* data = reinterpret_cast<uint8_t const*>(&marker);
+    auto* ptr = data + sizeof(MMFilesMarker);
+    auto* end = data + marker.getSize();
+
+    TRI_ASSERT(sizeof(TRI_voc_tick_t) <= size_t(end - ptr));
+    _databaseId = arangodb::encoding::readNumber<TRI_voc_tick_t>(
+      ptr, sizeof(TRI_voc_tick_t)
+    );
+    ptr += sizeof(TRI_voc_tick_t);
+    _slice = arangodb::velocypack::Slice(ptr);
+    TRI_ASSERT(_slice.byteSize() == size_t(end - ptr));
+  }
+
+  /// @brief write constructor
+  MMFilesFlushMarker(
+      TRI_voc_tick_t databaseId,
+      arangodb::velocypack::Slice const& slice
+  ): _databaseId(databaseId), _slice(slice) {
+  }
+
+  TRI_voc_tick_t databaseId() const noexcept { return _databaseId; }
+
+  // always 0 since not yet in any datafile
+  TRI_voc_fid_t fid() const override final { return 0; }
```

Comment:
why is this needed?

Reply:
databaseId() because _databaseId member is private
fid() because MMFilesWalMarker::fid() is abstract